canvas_sdk/data/data_access_layer_client.py

```diff
@@ -0,0 +1,87 @@
+"""
+Data Access Layer client.
+
+This module is primarily responsible for executing calls to the gRPC service so that such details
+are abstracted away from callers. The return values of the methods on the client class are protobufs
+which must be mapped to user-facing objects.
+"""
+
+import functools
+from collections.abc import Callable
+from types import FunctionType
+from typing import Any
+
+import grpc
+from grpc import StatusCode
+
+from canvas_generated.data_access_layer.data_access_layer_pb2 import ID, Patient
+from canvas_generated.data_access_layer.data_access_layer_pb2_grpc import (
+    DataAccessLayerStub,
+)
+from settings import DAL_TARGET
+
+from . import exceptions
+
+
+class _DataAccessLayerClientMeta(type):
+    """
+    Metaclass for the Data Access Layer client class.
+
+    Wraps all methods of a class with a gRPC error handler.
+    """
+
+    def __new__(cls, name: str, bases: tuple, attrs: dict) -> type:
+        for name, value in attrs.items():
+            if isinstance(value, FunctionType):
+                attrs[name] = cls.handle_grpc_errors(value)
+        return super().__new__(cls, name, bases, attrs)
+
+    @classmethod
+    def handle_grpc_errors(cls, func: Callable[..., Any]) -> Callable[..., Any]:
+        """
+        Decorator that wraps a try-except block around all class methods. gRPC errors are mapped to
+        a defined set of exceptions from a Data Access Layer exception hierarchy.
+        """
+
+        @functools.wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            try:
+                return func(*args, **kwargs)
+            except grpc.RpcError as error:
+                # gRPC exceptions aren't tightly defined, so we'll try to get a status code and
+                # handle it if we don't
+                try:
```

Comment:
Might need a separate try-except to get `details` as well, since it's not guaranteed to be there.